bionic/core/flow_execution.py

```diff
@@ -104,39 +110,44 @@ def _process_active_entry(self, entry):
 
         # In theory we could have an entry that doesn't require priming, but in practice
         # we always require it when we create one.
-        assert entry.is_requested_to_be_primed
+        assert entry.required_level in (EntryLevel.PRIMED, EntryLevel.CACHED)
 
-        # First, if we already have a cached value, we can stop immediately without
+        # First, if we've already met our requirements, we can stop immediately without
         # needing to process any dependencies.
-        if entry.is_cached:
+        if entry.all_incoming_reqs_are_met:
             self._mark_entry_completed(entry)
             return
 
-        # Otherwise, we'll need all of our dependencies to be primed so that their
-        # digests are available.
+        # Otherwise, we can't do anything more without the digests of our dependencies,
+        # so we'll need them to be primed.
         self._set_up_entry_dependencies(entry)
         for dep_entry in entry.dep_entries:
-            self._request_entry_be_primed(dep_entry)
-        if not entry.all_dependencies_satisfy_requests:
+            self._add_requirement(entry, dep_entry, EntryLevel.PRIMED)
+        if not entry.all_outgoing_reqs_are_met:
             self._mark_entry_blocked(entry)
             return
 
+        # Now that ew have the dependency digests, we can initialize our task state.
         entry.state.initialize(self._bootstrap, self._flow_instance_uuid)
 
-        # If this entry is persistable, let's see if we can load it from the cache ...
+        # If this entry is persistable, we may be able to load it from the persistent
+        # cache, which will immediately get us to the CACHED level.
         if entry.state.should_persist:
             entry.state.attempt_to_access_persistent_cached_values()
             if entry.is_cached:
                 self._mark_entry_completed(entry)
                 return
+
+        # Otherwise, if it's not persistable, then initializing it will have gotten it
+        # to the PRIMED level. If that's all we needed, we're done.
         else:
-            # ... otherwise, if the digest was all we needed, calling initialize() was
-            # sufficient.
-            if not entry.is_requested_to_be_cached:
+            assert entry.level >= EntryLevel.PRIMED
```

Comment:
There is a similar assert at line 113 that checks level in a list of acceptable levels. I personally like this way better (using `>=` operator) over the list way (using the `in` operator on the list), but I'm fine with either approach. Can you pick one and make it consistent?

Reply:
I originally used `>=` in both places, but in the context of the first assert it seemed helpful to make it clear that there are only two levels of interest, since there are a couple places where we assume that being cached is the same as being done. However, your comment on `is_cached` removes one of them, so maybe if we add an `assert entry.all_incoming_reqs_are_complete` before the final `_mark_entry_completed`, that assumption isn't necessary anymore.